New arguments, new columns, and rename columns about `*isic_4digit*`

kalashsinghal opened the pull request:
The `isic_tilt_mapper` file help us to map the name of the `isic_4digit` sectors to a separate column named `isic_4digit_name`.

@maurolepore As per the new requirement, `isic_4digit_name` column needs to be added to sector profile indicators as well which was already beed added to Emissions profile indicators. 

---

This reprex shows that `profile_sector*()` at product level now output `*isic_4digit*` columns. 

``` r
devtools::load_all()
#> ℹ Loading tiltIndicatorAfter
library(tiltToyData)
library(dplyr, warn.conflicts = FALSE)
library(readr, warn.conflicts = FALSE)

options(readr.show_col_types = FALSE)

companies <- read_csv(toy_sector_profile_companies())
scenarios <- read_csv(toy_sector_profile_any_scenarios())

result <- profile_sector(
  companies,
  scenarios,
  # TODO: Move to tiltToyData
  europages_companies = ep_companies |> head(3),
  ecoinvent_activities = ecoinvent_activities |> head(3),
  ecoinvent_europages = matches_mapper |> head(3),
  isic_tilt = isic_tilt_mapper |> head(3)
)

result |> unnest_product() |> relocate(matches("isic"))
#> # A tibble: 196 × 28
#>    isic_4digit isic_4digit_name companies_id                company_name country
#>    <chr>       <chr>            <chr>                       <chr>        <chr>  
#>  1 2410        <NA>             fleischerei-stiefsohn_0000… <NA>         <NA>   
#>  2 2410        <NA>             fleischerei-stiefsohn_0000… <NA>         <NA>   
#>  3 2410        <NA>             fleischerei-stiefsohn_0000… <NA>         <NA>   
#>  4 2410        <NA>             fleischerei-stiefsohn_0000… <NA>         <NA>   
#>  5 2410        <NA>             fleischerei-stiefsohn_0000… <NA>         <NA>   
#>  6 2410        <NA>             fleischerei-stiefsohn_0000… <NA>         <NA>   
#>  7 2410        <NA>             fleischerei-stiefsohn_0000… <NA>         <NA>   
#>  8 2410        <NA>             fleischerei-stiefsohn_0000… <NA>         <NA>   
#>  9 2410        <NA>             fleischerei-stiefsohn_0000… <NA>         <NA>   
#> 10 2410        <NA>             fleischerei-stiefsohn_0000… <NA>         <NA>   
#> # ℹ 186 more rows
#> # ℹ 23 more variables: PSTR_risk_category <chr>, scenario <chr>, year <dbl>,
#> #   ep_product <chr>, matched_activity_name <chr>,
#> #   matched_reference_product <chr>, unit <chr>, tilt_sector <chr>,
#> #   tilt_subsector <chr>, multi_match <lgl>, matching_certainty <chr>,
#> #   matching_certainty_company_average <chr>, company_city <chr>,
#> #   postcode <dbl>, address <chr>, main_activity <chr>, …

result |> unnest_company() |> relocate(matches("isic"))
#> # A tibble: 588 × 12
#>    companies_id      company_name country PSTR_share PSTR_risk_category scenario
#>    <chr>             <chr>        <chr>        <dbl> <chr>              <chr>   
#>  1 fleischerei-stie… <NA>         <NA>             1 high               IPR 1.5…
#>  2 fleischerei-stie… <NA>         <NA>             0 medium             IPR 1.5…
#>  3 fleischerei-stie… <NA>         <NA>             0 low                IPR 1.5…
#>  4 fleischerei-stie… <NA>         <NA>             1 high               IPR 1.5…
#>  5 fleischerei-stie… <NA>         <NA>             0 medium             IPR 1.5…
#>  6 fleischerei-stie… <NA>         <NA>             0 low                IPR 1.5…
#>  7 fleischerei-stie… <NA>         <NA>             0 high               announc…
#>  8 fleischerei-stie… <NA>         <NA>             0 medium             announc…
#>  9 fleischerei-stie… <NA>         <NA>             1 low                announc…
#> 10 fleischerei-stie… <NA>         <NA>             0 high               announc…
#> # ℹ 578 more rows
#> # ℹ 6 more variables: year <chr>, matching_certainty_company_average <chr>,
#> #   company_city <chr>, postcode <dbl>, address <chr>, main_activity <chr>



companies <- read_csv(toy_sector_profile_upstream_companies())
scenarios <- read_csv(toy_sector_profile_any_scenarios())
inputs <- read_csv(toy_sector_profile_upstream_products())
ecoinvent_inputs <- ecoinvent_inputs |> head(3)

result <- profile_sector_upstream(
  companies,
  scenarios,
  inputs,
  # TODO: Move to tiltToyData
  europages_companies = ep_companies |> head(3),
  ecoinvent_activities = ecoinvent_activities |> head(3),
  ecoinvent_inputs = ecoinvent_inputs |> head(3),
  ecoinvent_europages = matches_mapper |> head(3),
  isic_tilt = isic_tilt_mapper |> head(3)
)

result |> unnest_product() |> relocate(matches("isic"))
#> # A tibble: 704 × 31
#>    input_isic_4digit input_isic_4digit_name companies_id    company_name country
#>    <chr>             <chr>                  <chr>           <chr>        <chr>  
#>  1 3821              <NA>                   fleischerei-st… <NA>         <NA>   
#>  2 3821              <NA>                   fleischerei-st… <NA>         <NA>   
#>  3 3821              <NA>                   fleischerei-st… <NA>         <NA>   
#>  4 3821              <NA>                   fleischerei-st… <NA>         <NA>   
#>  5 3821              <NA>                   fleischerei-st… <NA>         <NA>   
#>  6 3821              <NA>                   fleischerei-st… <NA>         <NA>   
#>  7 3821              <NA>                   fleischerei-st… <NA>         <NA>   
#>  8 3821              <NA>                   fleischerei-st… <NA>         <NA>   
#>  9 3821              <NA>                   fleischerei-st… <NA>         <NA>   
#> 10 3821              <NA>                   fleischerei-st… <NA>         <NA>   
#> # ℹ 694 more rows
#> # ℹ 26 more variables: ISTR_risk_category <chr>, scenario <chr>, year <dbl>,
#> #   ep_product <chr>, matched_activity_name <chr>,
#> #   matched_reference_product <chr>, unit <chr>, tilt_sector <chr>,
#> #   multi_match <lgl>, matching_certainty <chr>,
#> #   matching_certainty_company_average <chr>, input_name <chr>,
#> #   input_unit <chr>, input_tilt_sector <chr>, input_tilt_subsector <chr>, …

result |> unnest_company() |> relocate(matches("isic"))
#> # A tibble: 294 × 12
#>    companies_id      company_name country ISTR_share ISTR_risk_category scenario
#>    <chr>             <chr>        <chr>        <dbl> <chr>              <chr>   
#>  1 fleischerei-stie… <NA>         <NA>        0.333  high               IPR 1.5…
#>  2 fleischerei-stie… <NA>         <NA>        0.583  medium             IPR 1.5…
#>  3 fleischerei-stie… <NA>         <NA>        0.0833 low                IPR 1.5…
#>  4 fleischerei-stie… <NA>         <NA>        1      high               IPR 1.5…
#>  5 fleischerei-stie… <NA>         <NA>        0      medium             IPR 1.5…
#>  6 fleischerei-stie… <NA>         <NA>        0      low                IPR 1.5…
#>  7 fleischerei-stie… <NA>         <NA>        0      high               announc…
#>  8 fleischerei-stie… <NA>         <NA>        0      medium             announc…
#>  9 fleischerei-stie… <NA>         <NA>        1      low                announc…
#> 10 fleischerei-stie… <NA>         <NA>        0.0769 high               announc…
#> # ℹ 284 more rows
#> # ℹ 6 more variables: year <chr>, matching_certainty_company_average <chr>,
#> #   company_city <chr>, postcode <dbl>, address <chr>, main_activity <chr>
```

<sup>Created on 2023-12-04 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

kalashsinghal commented on R/prepare_istr_company.R:
```diff
@@ -7,29 +7,16 @@
 #'
 #' @export
-#'
-#' @examples
-#' library(tiltIndicator)
-#'
-#' company <- unnest_company(toy_sector_profile_upstream_output())
-#' product <- unnest_product(toy_sector_profile_upstream_output())
-#'
-#' prepare_istr_company(
-#'   company |> head(3),
-#'   product |> head(3),
-#'   ep_companies |> head(3),
-#'   ecoinvent_activities |> head(3),
-#'   matches_mapper |> head(3),
-#'   ecoinvent_inputs |> head(3)
```
@maurolepore I had to remove these examples from the code because the `toy_sector_profile_upstream_output()` dataset does not contain the `isic_4digit` column required for joining the `isic_4digit_name` column through the `isic_tilt_mapper` file. 

At the moment the removal of these examples makes sense because we do not have right toy datasets for tiltIndicatorAfter. Hence, I would like you to decide how to proceed further with this issue. Should I create a separate issue for creating the toy datasets after merging this code? or you would be more interested in creating the new toy datasets first and write correct examples and then push this code?

kalashsinghal commented on tests/testthat/test-prepare_istr_company.R:
```diff
@@ -1,23 +1,46 @@
 
 test_that("handles tiltIndicator output", {
-  company <- unnest_company(toy_sector_profile_upstream_output())
-  product <- unnest_product(toy_sector_profile_upstream_output())
+  local_options(readr.show_col_types = FALSE)
+
+  companies <- read_csv(toy_sector_profile_upstream_companies())
+  scenarios <- read_csv(toy_sector_profile_any_scenarios())
+  inputs <- read_csv(toy_sector_profile_upstream_products())
+  .inputs <- add_rowid(inputs)
+  output <- sector_profile_upstream(companies, scenarios, .inputs)
+  .product <- unnest_product(output)
+
+  y <- select(.inputs, matches(extra_cols_pattern()))
+  x <- select(.product, -any_of(names(y)), extra_rowid())
+  product <- left_join(x, y, relationship = "many-to-many", by = extra_rowid())
+  company <- unnest_company(output)
```
@maurolepore To run the old tests, I had to write extra code which needs to be fixed using correct toy datasets. (Flagging)

maurolepore replied:
See my previous comment about hacking `toy_sector_profile_upstream_output()` to add the column you want and point to the relevant issue/PR in tiltToyData. The one change in `toy_sector_profile_upstream_output()` would avoid many many changes in this PR.

maurolepore commented on R/prepare_ictr_product.R:
```diff
@@ -54,7 +54,7 @@ rename_ictr_product <- function(data) {
       input_name = "exchange_name",
       input_unit = "exchange_unit_name",
-      input_isic_name = "isic_4digit_name_ecoinvent"
+      isic_4digit_name = "isic_4digit_name_ecoinvent"
```
The title of the PR is misleading:

* This change affects `*ictr*()` (now `profile_emissions_upstream()`), so it is not limited to the `*sector*()` functions.
* This change renames a column, so it's not limited to adding a column.

Also, why do we remove the `input_` prefix on this particular column? Although I would love to remove that prefix I think we should either remove it from everywhere or  nowhere.

maurolepore reviewed (commented):
.

maurolepore commented on R/prepare_istr_product.R:
```diff
@@ -5,26 +5,20 @@
-#'   ecoinvent_inputs |> head(3)
-#' )
-prepare_istr_product <- function(istr_prod, comp, eco_activities, match_mapper, eco_inputs) {
+prepare_istr_product <- function(istr_prod, comp, eco_activities, match_mapper, eco_inputs, isic_tilt_map) {
```
SELF: Passed to `exec_profile()`?

maurolepore commented on R/prepare_pstr_company.R:
```diff
@@ -4,30 +4,18 @@
 #' @param eco_activities A dataframe like [ecoinvent_activities]
 #' @param pstr_prod A dataframe like [pstr_product]
 #' @param comp A dataframe like [ep_companies]
+#' @param isic_tilt_map A dataframe like [isic_tilt_mapper]
```
SELF: Passed to `exec*()`?

Documented in all functions?

maurolepore commented on R/prepare_pstr_company.R:
```diff
@@ -4,30 +4,18 @@
 #' @return A dataframe that prepares the final output of pstr_company
 #'
 #' @export
-#'
-#' @examples
-#' library(tiltIndicator)
-#'
-#' company <- unnest_company(toy_sector_profile_upstream_output())
-#' product <- unnest_product(toy_sector_profile_upstream_output())
-#'
-#' prepare_pstr_company(
-#'   company |> head(3),
-#'   product |> head(3),
-#'   ep_companies |> head(3),
-#'   ecoinvent_activities |> head(3),
-#'   matches_mapper |> head(3)
-#' )
-prepare_pstr_company <- function(pstr_comp, pstr_prod, comp, eco_activities, match_mapper) {
```
See my comment above. You don't need to remove the example -- instead hack `toy_sector_profile_upstream_output()` (which lives in this repo) and add a `# FIXME` pointing to an issue or PR in tiltToyData.

maurolepore commented on R/prepare_pstr_company.R:
```diff
@@ -4,30 +4,18 @@
-#'   matches_mapper |> head(3)
-#' )
-prepare_pstr_company <- function(pstr_comp, pstr_prod, comp, eco_activities, match_mapper) {
+prepare_pstr_company <- function(pstr_comp, pstr_prod, comp, eco_activities, match_mapper, isic_tilt_map) {
```
SELF: Passed to `exec()`?

maurolepore commented on R/profile_sector_upstream.R:
```diff
@@ -45,7 +46,8 @@
 #'   europages_companies = ep_companies |> head(3),
 #'   ecoinvent_activities = ecoinvent_activities |> head(3),
 #'   ecoinvent_europages = matches_mapper |> head(3),
-#'   ecoinvent_inputs = ecoinvent_inputs |> head(3)
+#'   ecoinvent_inputs = ecoinvent_inputs |> head(3),
+#'   isic_tilt = isic_tilt_mapper |> head(3)
```
SELF: Check order. `eocinvent_inputs` should go before `ecoinvent_europages`?

maurolepore commented on R/profile_sector_upstream.R:
```diff
@@ -58,6 +60,7 @@ profile_sector_upstream <- function(companies,
                                     ecoinvent_activities,
                                     ecoinvent_europages,
                                     ecoinvent_inputs,
+                                    isic_tilt,
```
SELF: Is this documented?

maurolepore replied:
NEWS:

* `profile_sector_upstream()` gains the argument `isic_tilt`.

maurolepore reviewed (changes requested):
.

maurolepore commented on R/prepare_pstr_company.R:
```diff
@@ -14,20 +15,22 @@
 #' library(tiltIndicator)
 #'
-#' company <- unnest_company(toy_sector_profile_upstream_output())
-#' product <- unnest_product(toy_sector_profile_upstream_output())
+#' company <- unnest_company(toy_sector_profile_output())
+#' product <- unnest_product(toy_sector_profile_output())
```
This was my mistake in the example.

maurolepore commented on R/toy_result.R:
```diff
@@ -65,5 +66,12 @@ toy_sector_profile_upstream_output <- function() {
+}
+
+# isic_4digit: https://github.com/2DegreesInvesting/tiltToyData/issues/15
+FIXME_mutate_product <- function(data, ...) {
+  data |>
+    mutate(product = lapply(product, \(x) mutate(x, ...)))
```
Here is the hack. It allows to prepare this PR with minimum change here while we work on tiltToyData#15

maurolepore reviewed (approved):
LGTM

Thanks @kalashsinghal 

* I pushed a commit that hacks the toy data to add a `*isic*` column you want so that we can work on this PR without having to remove tests and examples. We should update the toy data (2DegreesInvesting/tiltToyData#15) and remove the hack before we merge this PR. Which[ toy datasets](https://2degreesinvesting.github.io/tiltToyData/reference/index.html) need the change?

* I think the title of this PR is misleading: The changes I see here are not limited to the `*sector*()` functions and not limited to adding a column. I see this PR affects also the `*emissions*()` functions and also rename an existing column. 

* I see an inconsistency: We remove the prefix `input_*` from one column of the output of `*_upstream()` but we leave the prefix from all other columns.

* This is a breaking, user-facing change. When we merge this PR we need to remember to document the change in NEWS.md (I use the [fledge](https://fledge.cynkra.com/) package).

kalashsinghal commented:
> I think the title of this PR is misleading: The changes I see here are not limited to the *sector*() functions and not limited to adding a column. I see this PR affects also the *emissions*() functions and also rename an existing column.

> I see an inconsistency: We remove the prefix input_* from one column of the output of *_upstream() but we leave the prefix from all other columns.

@maurolepore I have changed the column names of `*_upstream()` indicators and also updated the title. I have done this to remain align at the moment, however these name changes should be homogenised across all three packages which is the task for later

maurolepore commented on R/prepare_istr_product.R:
```diff
@@ -46,7 +53,8 @@ rename_istr_product <- function(data) {
       ISTR_risk_category = "risk_category",
       input_name = "exchange_name",
-      input_unit = "exchange_unit_name"
+      input_unit = "exchange_unit_name",
+      input_isic_4digit_name = "isic_4digit_name_ecoinvent"
```
NEWS:

* The output of `profile_sector_upstream()` at product level gains the column `input_isic_4digit_name`.

maurolepore commented on R/prepare_pctr_product.R:
```diff
@@ -41,7 +41,7 @@ rename_pctr_product <- function(data) {
       PCTR_risk_category = "risk_category",
       ep_product = "clustered",
-      isic_name = "isic_4digit_name_ecoinvent"
+      isic_4digit_name = "isic_4digit_name_ecoinvent"
```
* In the output of `profile_emissions()` at product level the column `isic_name` is now renamed to `input_isic_4digit_name`.

kalashsinghal replied:
@maurolepore it's renamed to `isic_4digit_name` (Just mentioning the typo)

maurolepore commented on R/prepare_pstr_product.R:
```diff
@@ -32,7 +27,8 @@ rename_pstr_product <- function(data) {
       matching_certainty_company_average = "avg_matching_certainty",
       PSTR_risk_category = "risk_category",
-      ep_product = "clustered"
+      ep_product = "clustered",
+      isic_4digit_name = "isic_4digit_name_ecoinvent"
```
NEWS:

* The output of `profile_sector()` at product level gains the column `input_isic_4digit_name`.

kalashsinghal replied:
@maurolepore it's renamed to `isic_4digit_name` (Just mentioning the typo)

maurolepore commented on R/profile_sector.R:
```diff
@@ -16,7 +17,8 @@ profile_sector <- function(companies,
     europages_companies,
     ecoinvent_activities,
-    ecoinvent_europages
+    ecoinvent_europages,
+    isic_tilt
```
NEWS:

* `profile_sector()` gains the argument `isic_tilt`.

maurolepore reviewed (commented):
Here are the user-facing chagnes that we need to report in NEWS.
